features/steps/component.go

```diff
@@ -0,0 +1,235 @@
+	}
+
+	// consumer for receiving events
+	cgChannels := kafka.CreateConsumerGroupChannels(1)
+	kafkaOffset := kafka.OffsetNewest
+	if cfg.Kafka.OffsetOldest {
+		kafkaOffset = kafka.OffsetOldest
+	}
+
+	if c.consumer, err = kafka.NewConsumerGroup(
+		ctx,
+		cfg.Kafka.Addr,
+		cfg.Kafka.CantabularCSVWCreatedTopic,
+		"csvw-created-group",
+		cgChannels,
+		&kafka.ConsumerGroupConfig{
+			KafkaVersion: &cfg.Kafka.Version,
+			Offset:       &kafkaOffset,
+		},
+	); err != nil {
+		return fmt.Errorf("error creating kafka consumer: %w", err)
+	}
+
+	// start kafka logging go-routines
+	c.producer.Channels().LogErrors(ctx, "component producer")
+	c.consumer.Channels().LogErrors(ctx, "component consumer")
+
+	// wait for consumer and producer to be ready
+	<-c.consumer.Channels().Ready
+	log.Info(ctx, "consumer ready")
+
+	<-c.producer.Channels().Ready
+	log.Info(ctx, "producer ready")
+
+	// Create service and initialise it
+	c.svc = service.New()
+	if err = c.svc.Init(ctx, cfg, BuildTime, GitCommit, Version); err != nil {
+		return fmt.Errorf("failed to init service: %w", err)
+	}
+
+	c.cfg = cfg
+
+	return nil
+}
+
+func (c *Component) startService(ctx context.Context) {
+	defer c.wg.Done()
+
+	c.svc.Start(context.Background(), c.errorChan)
+	c.svcStarted <- true
+
+	// blocks until an os interrupt or a fatal error occurs
+	select {
+	case err := <-c.errorChan:
+		err = fmt.Errorf("service error received: %w", err)
+		c.svc.Close(ctx)
+		panic(fmt.Errorf("unexpected error received from errorChan: %w", err))
+	case sig := <-c.signals:
+		log.Info(ctx, "os signal received", log.Data{"signal": sig})
+	}
+	if err := c.svc.Close(ctx); err != nil {
+		panic(fmt.Errorf("unexpected error during service graceful shutdown: %w", err))
+	}
+}
+
+// drainTopic drains the topic of any residual messages between scenarios.
+// Prevents future tests failing if previous tests fail unexpectedly and
+// leave messages in the queue.
+func (c *Component) drainTopic(ctx context.Context) error {
+	var msgs []interface{}
+
+	defer func() {
+		log.Info(ctx, "drained topic", log.Data{
+			"len":      len(msgs),
+			"messages": msgs,
+		})
+	}()
+
+	for {
+		select {
+		case <-time.After(time.Second * 1):
+			return nil
+		case msg, ok := <-c.consumer.Channels().Upstream:
+			if !ok {
+				return errors.New("upstream channel closed")
+			}
+
+			msgs = append(msgs, msg)
+			msg.Commit()
+			msg.Release()
+		}
+	}
+}
+
+// Close kills the application under test, and then it shuts down the testing consumer and producer.
+func (c *Component) Close() {
+	ctx := context.Background()
+
+	if err := c.drainTopic(ctx); err != nil {
+		log.Error(ctx, "error draining topic", err)
+	}
+
+	// close producer
+	if err := c.producer.Close(ctx); err != nil {
+		log.Error(ctx, "error closing kafka producer", err)
+	}
+
+	// close consumer
+	if err := c.consumer.Close(ctx); err != nil {
+		log.Error(ctx, "error closing kafka consumer", err)
+	}
+
+	// kill application
```

Comment:
In other repos, I first killed the app and then the component tests (the logic was that once the service is dead, then we can do any clean-up without interfering with it), but it seems to be ok.

Reply:
Haven't had any issues so far